Instrumented clrjit causes random msbuild failures

`dotnet-optimization` pipeline is currently blocked on this issue, I managed to make a minimal standalone repro (Linux-x64):

Create a `repro.sh` script with the following content:
```bash
#!/bin/bash
set -e

rm -rf repro
#SDK_TYPE=-pgo

# Step 1: download the latest dotnet SDK

mkdir repro
cd repro

wget -O dotnet-sdk.tar.gz https://dotnetbuilds.azureedge.net/public/Sdk/7.0.100-preview.4.22204.1/dotnet-sdk$SDK_TYPE-7.0.100-preview.4.22204.1-linux-x64.tar.gz
mkdir dotnet-sdk && tar -xvzf dotnet-sdk.tar.gz -C dotnet-sdk

# Step 2: download OrchardCore project 

git clone https://github.com/dotnet-perf-bot/orchardcore.git
cd orchardcore/src/OrchardCore.Cms.Web

# Step 3: Try to build OrchardCore with instrumented SDK:
DOTNET_MULTILEVEL_LOOKUP=0 DOTNET_SKIP_FIRST_TIME_EXPERIENC=1 DOTNET_CLI_TELEMETRY_OPTOUT=1 ../../../dotnet-sdk/dotnet build -c Release
```
and run it - it should complete without errors.

Now uncomment `#SDK_TYPE=-pgo` line to use **natively** instrumented SDK and run again - observe random meaningless msbuild failures. e.g.
```
/home/egorbo/prj/repro/dotnet-sdk/sdk/7.0.100-preview.4.22204.1/Current/Microsoft.Common.props(29,5): error MSB4186: 
Invalid static method invocation syntax: "[MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildProjectDirectory), 
'$(_DirectoryBuildPropsFile)')". Method '[MSBuild]::GetDirectoryNameOfFileAbove' not found. Static method invocation should 
be of the form: $([FullTypeName]::Method()), e.g. $([System.IO.Path]::Combine(`a`, `b`)). Check that all parameters are defined,
 are of the correct type, and are specified in the right order. 
[/home/egorbo/prj/repro/orchardcore/src/OrchardCore/OrchardCore/OrchardCore.csproj]
```

After some experiments I came to conclusion that it's `libclrjit.so` binary that causes it - I downloaded both instrumented and normal sdk and only copied `libclrjit.so` from the pgo to normal one and removed pgo:
<details>
  <summary>Replace clrjit from pgo</summary>

```
#!/bin/bash
set -e

rm -rf repro
# Step 1: download the latest dotnet SDK

mkdir repro
cd repro

wget -O dotnet-sdk-pgo.tar.gz https://dotnetbuilds.azureedge.net/public/Sdk/7.0.100-preview.4.22204.1/dotnet-sdk-pgo-7.0.100-preview.4.22204.1-linux-x64.tar.gz
wget -O dotnet-sdk.tar.gz https://dotnetbuilds.azureedge.net/public/Sdk/7.0.100-preview.4.22204.1/dotnet-sdk-7.0.100-preview.4.22204.1-linux-x64.tar.gz
mkdir dotnet-sdk && tar -xvzf dotnet-sdk.tar.gz -C dotnet-sdk
mkdir dotnet-sdk-pgo && tar -xvzf dotnet-sdk-pgo.tar.gz -C dotnet-sdk-pgo

cp -f dotnet-sdk-pgo/shared/Microsoft.NETCore.App/7.0.0-preview.4.22201.3/libclrjit.so dotnet-sdk/shared/Microsoft.NETCore.App/7.0.0-preview.4.22201.3/libclrjit.so
rm -rf dotnet-sdk-pgo

# Step 2: download OrchardCore project 

git clone https://github.com/dotnet-perf-bot/orchardcore.git
cd orchardcore/src/OrchardCore.Cms.Web

# Step 3: Try to build OrchardCore with instrumented SDK:
DOTNET_MULTILEVEL_LOOKUP=0 DOTNET_SKIP_FIRST_TIME_EXPERIENC=1 DOTNET_CLI_TELEMETRY_OPTOUT=1 ../../../dotnet-sdk/dotnet build -c Release orchardcore/src/OrchardCore.Cms.Web

```
</details>

cc @dotnet/jit-contrib @jkotas @davidwrighton while I try to diagnose it further maybe any advices what to check - can instrumentation inside the jit cause gc holes?

I tried to run with non-instrumented sdk and GCStress=0xC - no issues

--- a/src/coreclr/jit/importer_vectorization.cpp
+++ b/src/coreclr/jit/importer_vectorization.cpp
@@ -88,17 +88,23 @@ static GenTreeHWIntrinsic* CreateConstVector(Compiler* comp, var_types simdType,
 #ifdef TARGET_XARCH
     if (simdType == TYP_SIMD32)
     {
-        GenTree* long1 = comp->gtNewIconNode(*(ssize_t*)(cns + 0), TYP_LONG);
-        GenTree* long2 = comp->gtNewIconNode(*(ssize_t*)(cns + 4), TYP_LONG);
-        GenTree* long3 = comp->gtNewIconNode(*(ssize_t*)(cns + 8), TYP_LONG);
-        GenTree* long4 = comp->gtNewIconNode(*(ssize_t*)(cns + 12), TYP_LONG);
+        ssize_t fourLongs[4];
+        CopyMemory(fourLongs, cns, sizeof(ssize_t) * 4);
+
+        GenTree* long1 = comp->gtNewIconNode(fourLongs[0], TYP_LONG);
+        GenTree* long2 = comp->gtNewIconNode(fourLongs[1], TYP_LONG);
+        GenTree* long3 = comp->gtNewIconNode(fourLongs[2], TYP_LONG);
+        GenTree* long4 = comp->gtNewIconNode(fourLongs[3], TYP_LONG);
         return comp->gtNewSimdHWIntrinsicNode(simdType, long1, long2, long3, long4, NI_Vector256_Create, baseType, 32);
     }
 #endif // TARGET_XARCH
 
+    ssize_t twoLongs[2];
+    CopyMemory(twoLongs, cns, sizeof(ssize_t) * 2);
+
     assert(simdType == TYP_SIMD16);
-    GenTree* long1 = comp->gtNewIconNode(*(ssize_t*)(cns + 0), TYP_LONG);
-    GenTree* long2 = comp->gtNewIconNode(*(ssize_t*)(cns + 4), TYP_LONG);
+    GenTree* long1 = comp->gtNewIconNode(twoLongs[0], TYP_LONG);
+    GenTree* long2 = comp->gtNewIconNode(twoLongs[1], TYP_LONG);
     return comp->gtNewSimdHWIntrinsicNode(simdType, long1, long2, NI_Vector128_Create, baseType, 16);
 }
 

--- a/src/coreclr/jit/importer_vectorization.cpp
+++ b/src/coreclr/jit/importer_vectorization.cpp
@@ -89,7 +89,7 @@ static GenTreeHWIntrinsic* CreateConstVector(Compiler* comp, var_types simdType,
     if (simdType == TYP_SIMD32)
     {
         ssize_t fourLongs[4];
-        CopyMemory(fourLongs, cns, sizeof(ssize_t) * 4);
+        memcpy(fourLongs, cns, sizeof(ssize_t) * 4);
 
         GenTree* long1 = comp->gtNewIconNode(fourLongs[0], TYP_LONG);
         GenTree* long2 = comp->gtNewIconNode(fourLongs[1], TYP_LONG);
@@ -100,7 +100,7 @@ static GenTreeHWIntrinsic* CreateConstVector(Compiler* comp, var_types simdType,
 #endif // TARGET_XARCH
 
     ssize_t twoLongs[2];
-    CopyMemory(twoLongs, cns, sizeof(ssize_t) * 2);
+    memcpy(twoLongs, cns, sizeof(ssize_t) * 2);
 
     assert(simdType == TYP_SIMD16);
     GenTree* long1 = comp->gtNewIconNode(twoLongs[0], TYP_LONG);
@@ -173,7 +173,7 @@ GenTree* Compiler::impExpandHalfConstEqualsSIMD(
     WCHAR cnsValue[maxPossibleLength]    = {};
     WCHAR toLowerMask[maxPossibleLength] = {};
 
-    CopyMemory((UINT8*)cnsValue, (UINT8*)cns, len * sizeof(WCHAR));
+    memcpy((UINT8*)cnsValue, (UINT8*)cns, len * sizeof(WCHAR));
 
     if ((cmpMode == OrdinalIgnoreCase) && !ConvertToLowerCase(cnsValue, toLowerMask, len))
     {

